### DISCUSSION TITLE
[API Proposal]: Request to create ChainPreviousMemberNameAttribute like CallerMemberNameAttribute

### DISCUSSION THREAD

**Discussion by vit-h-zz** [dotnet/csharplang#5118] [Language Ideas]
### Background and motivation

I am wrapping the Task by extension method in detailed logging and missing a way to automatically get the name of the wrapped method.
Would be nice to add this tiny feature.

public static async Task<T> WithLogging<T>(this Task<T> task,
            **[ChainPreviousMemberName] string wrapThe = "",**
            [CallerMemberName] string method = "",
            [CallerFilePath] string srcFilePath = "",
            [CallerLineNumber] int srcLineNumber = 0)
            { ... }

### API Proposal

same as CallerMemberNameAttribute

### API Usage

public static async Task<T> WithLogging<T>(this Task<T> task,
            **[ChainPreviousMemberName] string wrapThe = "",**
            [CallerMemberName] string method = "",
            [CallerFilePath] string srcFilePath = "",
            [CallerLineNumber] int srcLineNumber = 0)
            { ... }

....

Then instead of this
await someService.AddOrUpdateAsync(something).WithLogging(nameof(AddOrUpdateAsync));

I can use this:
await someService.AddOrUpdateAsync(something).WithLogging();

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by Wraith2** [comment]
`CallerMemberName` functionality is implemented by the compiler and embeds the name of the caller into the callsites as a literal value. If you want to chain those calls you must control the methods that need to pass on the information. If you do not control the parameter list of one of the calling methods than that is the point that you cannot pass with this functionality. At that point your only option is to use StackTrace which is expensive.

**Reply by vit-h-zz** [reply]
@Wraith2 Thanks for your answer, but I did not fully get what is limiting such a feature.

Sure I want `[ChainPreviousMemberName] string prevChainMemberName` to be populated on the compile time.
So here is an example of the pain point which is in a million places in the app but with such feature may be automatically added with no additional symbols like `[CallerMemberName] string method`.

using System;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;


namespace ConsoleApp1
{

    class Program
    {
        static async void Main(string[] args)
        {
            var someServicLongName = new SomeService();
            var someData = new object();

            await someServicLongName
                .SaveSomethingAsync(someData)
                .WithLogging();

            await someServicLongName
                .SaveSomethingAsync(someData)
                .WithLogging(nameof(someServicLongName.SaveSomethingAsync));

            Console.ReadLine();
        }
    }

    static class Ext
    {
        public static async Task WithLogging(this Task task,
            [ChainPreviousMemberName] string prevChainMemberName = "",
            [CallerMemberName] string method = "")
        {
            try
            {
                Console.WriteLine(new { status = "Start", method, prevChainMemberName });

                await task;

                Console.WriteLine(new { status = "Done", method, prevChainMemberName });
            }
            catch (Exception ex)
            {
                Console.WriteLine(new { status = "Error: " + ex.Message, method, prevChainMemberName });
            }
        }
    }

    class SomeService
    {
        public Task SaveSomethingAsync(object data)
        {
            return Task.CompletedTask;
        }
    }
}


CC: @jkoritzinsky

**Comment by Wraith2** [comment]
> Thanks for your answer, but I did not fully get what is limiting such a feature.

If implemented using the current attribute It can never cross a method which does not have a `[CallerMemberName]` parameter. Consider:

```csharp
public class C {
    
    public static void Top1() => Middle();
    
    public static void Top2() => Middle();
    
    public static void Middle() => Bottom();
    
    public static void Bottom([CallerMemberName] string caller = null, string previousCaller = null)
    {
        Console.WriteLine($"{previousCaller}->{caller}->Bottom");
    }
}
```

If you call Top1 you'd expect to get the output `Top1->Middle->Bottom` and if you called Top2 get `Top2->Middle->Bottom` but how does middle pass the previous caller name to Bottom? Middle doesn't have that name because it has no caller name parameter. So you compile a different version of Bottom for each Top* method? that's pretty inefficient and will scale horribly, we haven't even started considering virtual invocations reflection emit etc.

Then you'll say that you can just add the `[CallerMemberName]` to the method. What if you can't? What happens if you're invoking it through a delegate or you're using a class derived from one in an external library where you can't change the public surface area?

In order to have this feature work the entire call stack would have to be co-operating and using `[CallerMemberName]` and if you are in that position you can add an extra parameter to pass through the parent caller name yourself without any additional work from the compiler. If you aren't in that position the compiler can't do magic to enable it.

**Reply by vit-h-zz** [reply]
Seems you are answering another question =)

I just want this feature

instead of writing this:
await **someServicLongName.SaveSomethingAsync**(someData)
.WithLogging(**nameof(someServicLongName.SaveSomethingAsync)**);

I want to write this:
await **someServicLongName.SaveSomethingAsync**(someData)
.WithLogging();

I want only the name **SaveSomethingAsync** from the someServicLongName.SaveSomethingAsync(someData) to be taken automatically by the new attribute.

so the WithLogging will look like:
public static async Task WithLogging(this Task task, [ChainPreviousMemberName] string wrappingMethod = "", ..... )
{
 Console.WriteLine("we are wrapping the " + wrappingMethod + " method with logging");
}

// Output: "we are wrapping the **SaveSomethingAsync** method with logging"

@Wraith2 Did I miss you or you were confused?
I think this is not a hard task to implement such a feature.

**Comment by Wraith2** [comment]
> @Wraith2 Did I miss you or you were confused?

I'm not confused. I understand what you want and I'm talking about how it would have to be implemented. Implementing the feature is important, you can't just cite magic.

> I think this is not a hard task to implement such a feature.

I think you're wrong.

**Reply by yaakov-h** [reply]
I do believe you are actually confused.

The OP does not want the name of the method on the previous stack frame, but the name of the method (IdentifierName) used on the left hand side of the current invocation expression that was used to call the method.
